llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -12416,7 +12416,7 @@ SDValue SITargetLowering::performRcpCombine(SDNode *N,
 }
 
 bool SITargetLowering::isCanonicalized(SelectionDAG &DAG, SDValue Op,
-                                       unsigned MaxDepth) const {
+                                       bool &Trunc, unsigned MaxDepth) const {
```

Comment:
If you always introduce a freeze whenever eliding any canonicalize, shouldn't need this special casing of trunc anymore?

Reply:
You're right, we can do that, it results in correct code, but that causes regressions where we are no longer able to figure out that we have an already canonicalised value and redundantly recanonicalise. I see that happen in the existing `clamp.ll`, `fmed3.ll`, `known-never-snan.ll`, `mad-mix-lo.ll`, and `minmax.ll` tests.

Reply:
With this different approach based on your idea, this is no longer needed.